### PR TITLE
EEG-1200A V01.00

### LINKED ISSUE #9462: cannnot read nihon kohden .EEG file by mne python
#### Describe the bug
I recently got a .eeg file from a hospital, but I cannot read it by MNE-python. I noticed that the version (EEG-1200A V01.00) is not supported at this moment.
Can anyone help with this problem?

#### Steps to reproduce

```python
import matplotlib
import pathlib
import mne

sample_data_dir = mne.datasets.sample.data_path()
raw_path = sample_data_dir / 'MEG' / 'sample' / 'FA0015AA.EEG'
raw = mne.io.read_raw_nihon(raw_path)
print(raw.info)
raw.plot()
```

#### Expected results

The .EEG data should be loaded.

#### Actual results

```
Loading FA0015AA.EEG
Reading header from C:\Users\Surface\mne_data\MNE-sample-data\MEG\sample\FA0015AA.EEG
Traceback (most recent call last):
  File "C:\Users\Surface\PycharmProjects\mne_plot\plot_test.py", line 13, in <module>
    raw = mne.io.read_raw_nihon(raw_path)
  File "C:\Users\Surface\PycharmProjects\mne_plot\venv\lib\site-packages\mne\io\nihon\nihon.py", line 46, in read_raw_nihon
    return RawNihon(fname, preload, verbose)
  File "<decorator-gen-243>", line 24, in __init__
  File "C:\Users\Surface\PycharmProjects\mne_plot\venv\lib\site-packages\mne\io\nihon\nihon.py", line 316, in __init__
    header = _read_nihon_header(fname)
  File "C:\Users\Surface\PycharmProjects\mne_plot\venv\lib\site-packages\mne\io\nihon\nihon.py", line 136, in _read_nihon_header
    raise ValueError(
ValueError: Not a valid Nihon Kohden EEG file (EEG-1200A V01.00)

Process finished with exit code 1
```

#### Additional information

- Platform:      Windows-10-10.0.19041-SP0
- Python:        3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)]
- Executable:    c:\users\surface\anaconda3\python.exe
- CPU:           Intel64 Family 6 Model 126 Stepping 5, GenuineIntel: 8 cores
- Memory:        15.6 GB

- mne:           0.23.0
- numpy:         1.20.1 {blas=mkl_rt, lapack=mkl_rt}
- scipy:         1.6.2
- matplotlib:    3.3.4 {backend=Qt5Agg}

- sklearn:       0.24.1
- numba:         0.53.1
- nibabel:       3.2.1
- nilearn:       Not found
- dipy:          Not found
- cupy:          Not found
- pandas:        1.2.4
- mayavi:        4.7.3
- pyvista:       Not found
- vtk:           9.0.1
- PyQt5:         5.15.4

### PR DESCRIPTION
**Pull request by jacobshaw42**
#### Reference issue
Fixes #9462.

#### What does this implement/fix?
Adds compatibility for Nihon EEG-1200A V01.00 format. This involves bypassing the set annotations method for only this version. 

#### Additional information
There is also a "sub-log" in the ```.LOG``` file. This adds more information to the annotations description and milliseconds to the annotation onsets.

Part of the "sub-log" code was inspired by [brainstorm](https://github.com/brainstorm-tools/brainstorm3/blob/master/toolbox/io/in_fopen_nk.m). Due to the differences in the flow of these functions, I did not "copy" or directly translate. However, I am not sure how the licenses play in this case. If this inspiration is not compatible with licenses, please disregard my pull request.

### CODE REVIEW

**Comment by drammock** [comment]
marking as draft based on [this comment](https://github.com/mne-tools/mne-python/issues/9462#issuecomment-1402240709):

> I did get some inspiration from the one of the brainstorm MATLAB functions.

@ftadel we need permission to relicense an adaptation of some Brainstorm code (GPL-3 to BST); it's for reading Nihon-Koden files.  OK for you?

**Comment by ftadel** [comment]
> @ftadel we need permission to relicense an adaptation of some Brainstorm code (GPL-3 to BST); it's for reading Nihon-Koden files. OK for you?

Sure. Please share this collective work as much as possible.

If you are implementing support for files that are currently not supported in Brainstorm (like EEG-1200A V01.00 with multiple data blocks), could you please let me know and share an example file with me? - Thanks

File: mne/io/nihon/nihon.py [hunk 1]
```diff
@@ -272,10 +274,19 @@ def _read_nihon_annotations(fname):
             n_logs = np.fromfile(fid, np.uint8, 1)[0]
             fid.seek(t_blk_address + 0x14)
             t_logs = np.fromfile(fid, '|S45', n_logs)
-            for t_log in t_logs:
+            if version == _valid_headers[-1]:
```
Comment (drammock) [inline mne/io/nihon/nihon.py:277]: The check for `if version == _valid_headers[-1]` happens a few times. It would be better not to rely on `_valid_headers` having a specific order, so I suggest a convenience variable, maybe like this

```py
is_version_1200A = version == 'EEG-1200A V01.00'  # add around line 265 or so
```

Not sure if it's the best naming choice since I'm unfamiliar with what this specific header is used for (a specific MEG device model?)

then this (and the other if-lines) would become 
```suggestion
            if is_version_1200A:
```

File: mne/io/nihon/nihon.py [hunk 2]
```diff
@@ -272,10 +274,19 @@ def _read_nihon_annotations(fname):
...
+                fid.seek(0x92 + (t_block + 22) * 20)
+                t_sub_blk_address = np.fromfile(fid,np.uint32,1)[0]
```
Comment (drammock) [inline mne/io/nihon/nihon.py:279]:
```suggestion
                t_sub_blk_address = np.fromfile(fid, np.uint32, 1)[0]
```

Style error. please run `make flake` from the root folder of the MNE clone to catch all the other ones, and fix them.

File: mne/io/nihon/nihon.py [hunk 3]
```diff
@@ -284,9 +295,15 @@ def _read_nihon_annotations(fname):
                     warn(f'Could not decode log as one of {_encodings}')
                     continue
                 t_desc = t_log[:20].strip('\x00')
+                if version == _valid_headers[-1]:
+                    t_sub_desc = t_sub_log[:20].strip('\x00')
+                    t_desc = t_desc + t_sub_desc
```
Comment (drammock) [inline mne/io/nihon/nihon.py:300]: Does it make sense to concatentate these with no separator? I would have expected maybe `' '.join([t_desc, t_sub_desc])` or `f'{t_desc} - {t_sub_desc}'` or something similar, but that's just a guess based on the variable names.

File: mne/io/nihon/nihon.py [hunk 4]
```diff
@@ -409,7 +426,10 @@ def __init__(self, fname, preload=False, verbose=None):
                 annots['description'].append('EDGE boundary')
 
         annotations = Annotations(**annots, orig_time=info['meas_date'])
-        self.set_annotations(annotations)
+        if header['version'] == _valid_headers[-1]:
+            self._annotations = annotations
+        else:
+            self.set_annotations(annotations)
```
Comment (drammock) [inline mne/io/nihon/nihon.py:432]: I'm still a bit wary of this change and would like input from other devs. `set_annotations` has some sensible sanity/safety checks and I don't like the idea of skipping them; I'd rather figure out why things aren't working and fix them.

**Review by drammock — COMMENTED** [review]
here's a first pass review @jacobshaw42.  Thanks for implementing this!  Ping us when these are addressed and you're ready for another look.  

Also can you share a file with this version that we can use for testing?  Apologies if you've posted the link elsewhere already and I just missed it.

Ideally we would (also) have a very short (just a few seconds) file to use in I/O unit tests. If there's a way to have it include fewer channels and still be useful for testing, that is also desirable.

File: mne/io/nihon/nihon.py (hunk 3 above)
Comment (jacobshaw42) [reply mne/io/nihon/nihon.py:300]: From my results, it does seem appropriate to concatenate with no separator.

File: mne/io/nihon/nihon.py (hunk 4 above)
Comment (jacobshaw42) [reply mne/io/nihon/nihon.py:433]: After looking into this more closely. I believe this has to do with the multiple blocks of data. There is only one collected (due to the version), so the rest of the annotations would be seen as out of range for the data found. Skipping this would allow to get the annotations even though we have no data for those annotations.

I am not sure that is helpful to have the annotations without the observations. I will spend sometime trying to get all of those.

**Comment by proloyd** [comment]
@drammock @jacobshaw42 did we discard the plan for supporting 'Nihon Kohden EEG-1200A V01.00'?

**Comment by drammock** [comment]
> Is there going to be a solution for supporting 'Nihon Kohden EEG-1200A V01.00'?

The contributor who was working on this has deleted their fork of MNE-Python (and so presumably abandoned the effort). Without access to a sample data file (ideally a "normal" one and also a very short one for using in tests), none of our devs can pick up where they left off. So at this point we're waiting on either an example file (and one of our devs to find time), or a different contributor who has an example file and who has time now to get this over the finish line (including some basic tests).

**Comment by wxl-001** [comment]
Support for 'Nihon Kohden EEG-1200A V01.00' is urgently needed. Though Neuroworkbench can convert to edf type, it is unrealistic to manually convert a large number of eeg files into edf.

**Comment by larsoner** [comment]
@wxl-001 do you have (or could you get) some small test files (e.g., 1 sec of data?) for us to test?

**Comment by wxl-001** [comment]
> @wxl-001 do you have (or could you get) some small test files (e.g., 1 sec of data?) for us to test?

I'm very sorry that I don't have the permission to share the data with others. Wish you could solve this problem as soon as possible.

**Comment by larsoner** [comment]
Can't work on it without files :disappointed:

**Comment by sarvizargar** [comment]
> Can't work on it without files 😞

Hi @larsoner,
I could provide you the files, will send them to your email: larson.eric.d@gmail.com
